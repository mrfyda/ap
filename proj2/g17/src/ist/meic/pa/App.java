package ist.meic.pa;

public class App {

    public static void main(String[] args) {
        //TODO: do something here
    }

}
