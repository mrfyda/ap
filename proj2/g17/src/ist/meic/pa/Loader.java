package ist.meic.pa;

import javassist.CannotCompileException;
import javassist.ClassPool;
import javassist.NotFoundException;
import javassist.Translator;

public class Loader extends javassist.Loader {
    private ClassPool source;
    private Translator translator;

    public Loader() {
    }

    @Override
    public void setClassPool(ClassPool cp) {
        super.setClassPool(cp);
        source = cp;
    }

    @Override
    public void addTranslator(ClassPool cp, Translator t)
            throws NotFoundException, CannotCompileException {
        super.addTranslator(cp, t);
        source = cp;
        translator = t;
    }

    @Override
    protected Class delegateToParent(String classname)
            throws ClassNotFoundException {

        try {
            if (source != null) {
                if (translator != null) {
                    translator.onLoad(source, classname);
                }
            }
        } catch (NotFoundException e) {
            System.out.println("[NotFoundException] Class: " + classname + ", this ain't good!");
        } catch (CannotCompileException e) {
            System.out.println("[CannotCompileException] Class: " + classname + ", this ain't good!");
        }

        ClassLoader cl = getParent();
        if (cl != null) {
            return cl.loadClass(classname);
        } else {
            return findSystemClass(classname);
        }
    }
}
