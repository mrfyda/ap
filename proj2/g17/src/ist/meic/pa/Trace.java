package ist.meic.pa;

import java.util.List;

public class Trace {

    public static void print(Object object) {
        List<TraceStep> steps = TraceHistory.traceSteps.get(object);

        if (steps == null) {
            System.out.println("Tracing for " + object + " is nonexistent!");
        } else {
            System.out.println("Tracing for " + object);

            for (TraceStep step : steps) {
                System.out.println(step);
            }
        }
    }

}
