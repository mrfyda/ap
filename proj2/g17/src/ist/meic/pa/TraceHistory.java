package ist.meic.pa;

import java.util.ArrayList;
import java.util.IdentityHashMap;
import java.util.List;

public class TraceHistory {

    public static IdentityHashMap<Object, List<TraceStep>> traceSteps = new IdentityHashMap<Object, List<TraceStep>>();

    public static void putLTR(Object[] objects, String file, String method, int line) {
        for (Object object : objects) {
            put(object, "->", file, method, line);
        }
    }

    public static void putRTL(Object object, String file, String method, int line) {
        put(object, "<-", file, method, line);
    }

    public static void putFieldWrite(Object object, String file, String fieldName, int line) {
        put(object, "Wrote ", file, fieldName, line);
    }

    public static void putFieldRead(Object object, String file, String fieldName, int line) {
        put(object, "Read ", file, fieldName, line);
    }

    public static void putHandler(Object object, String file, String exceptioName, int line) {
        put(object, "Catched ", file, exceptioName, line);
    }

    public static void putCast(Object object, String file, String className, int line) {
        put(object, "Cast to", file, className, line);
    }

    public static void putInstanceOf(Object object, String file, String className, int line) {
        put(object, "Checked instance of", file, className, line);
    }

    public static void putNewArray(Object object, String file, String className, int line) {
        put(object, "New array of ", file, className, line);
    }

    private static void put(Object object, String direction, String file, String method, int line) {
        List<TraceStep> steps = traceSteps.get(object);

        if (steps == null) {
            steps = new ArrayList<TraceStep>();
        }

        TraceStep step = new TraceStep(direction, method, file, line);
        steps.add(step);
        traceSteps.put(object, steps);
    }

    public static void printTraceSteps(Object object) {
        List<TraceStep> steps = TraceHistory.traceSteps.get(object);

        if (steps == null) {
            System.out.println("Tracing for " + object + " is nonexistent!");
        } else {
            System.out.println("Tracing for " + object);

            for (TraceStep step : steps) {
                System.out.println(step);
            }
        }
    }
}

class TraceStep {
    private String direction;
    private String method;
    private String file;
    private Integer line;

    public TraceStep(String direction, String method, String file, Integer line) {
        this.direction = direction;
        this.file = file;
        this.method = method;
        this.line = line;
    }

    @Override
    public String toString() {
        return String.format("  %s %s on %s:%d", direction, method, file, line);
    }
}
