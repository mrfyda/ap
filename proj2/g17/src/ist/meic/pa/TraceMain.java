package ist.meic.pa;

import javassist.*;

public class TraceMain {

    private Translator translator;

    public TraceMain() {
        this.translator = new TraceTranslator();
    }

    public TraceMain(Translator translator) {
        this.translator = translator;
    }

    public void run(String[] args) {
        if (args.length < 1) {
            System.err.println("usage: ist.meic.pa.TraceVM <target application class> [<target application arguments>]");
            System.exit(1);
        }

        ClassPool pool = ClassPool.getDefault();
        Loader classLoader = new Loader();

        try {
            classLoader.addTranslator(pool, this.translator);
        } catch (NotFoundException e) {
            e.printStackTrace();
        } catch (CannotCompileException e) {
            e.printStackTrace();
        }

        try {
            String[] restArgs = new String[args.length - 1];
            System.arraycopy(args, 1, restArgs, 0, restArgs.length);

            classLoader.run(args[0], restArgs);
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
    }

}
