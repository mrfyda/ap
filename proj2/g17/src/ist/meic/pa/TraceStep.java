package ist.meic.pa;

public class TraceStep {
    public String file;
    public String method;
    public Integer line;

    public TraceStep(String file, String method, Integer line) {
        this.file = file;
        this.method = method;
        this.line = line;
    }

    @Override
    public String toString() {
        return method + " on " + file + ":" + line;
    }
}
