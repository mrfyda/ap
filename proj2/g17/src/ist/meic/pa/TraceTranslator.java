package ist.meic.pa;

import javassist.*;
import javassist.expr.ExprEditor;
import javassist.expr.MethodCall;
import javassist.expr.NewExpr;

public class TraceTranslator implements Translator {

    @Override
    public void start(ClassPool pool) throws NotFoundException, CannotCompileException {
    }

    @Override
    public void onLoad(ClassPool pool, String classname) throws NotFoundException, CannotCompileException {
        CtClass clazz = pool.get(classname);

        if (clazz.getSimpleName().equals("TraceHistory")) return;

        clazz.instrument(new TraceExprEditor());
    }
}

class TraceExprEditor extends ExprEditor {

    public void edit(MethodCall methodCall) throws CannotCompileException {
        try {
            CtMethod method = methodCall.getMethod();

            methodCall.replace(
                    String.format("{" +
                            "$_ = $proceed($$);" +
                            "String filename = \"%s\";" +
                            "String method = \"%s\";" +
                            "int line = %d;" +
                            "ist.meic.pa.TraceHistory.putLTR($args, filename, method, line);" +
                            "ist.meic.pa.TraceHistory.putRTL(($w)$_, filename, method, line);" +
                            "}", methodCall.getFileName(), method.getLongName(), methodCall.getLineNumber())
            );
        } catch (NotFoundException ignored) {
        }
    }

    public void edit(NewExpr newExpr) throws CannotCompileException {
        try {
            CtConstructor constructor = newExpr.getConstructor();

            newExpr.replace(
                    String.format("{" +
                            "$_ = $proceed($$);" +
                            "if ($_ != null && $_.toString().length() > 0 && $_.toString() != \"{}\") {" +
                            "ist.meic.pa.TraceHistory.putRTL($_, \"%s\", \"%s\", %d);" +
                            "}" +
                            "}", newExpr.getFileName(), constructor.getLongName(), newExpr.getLineNumber())
            );
        } catch (NotFoundException ignored) {
        }
    }

}
