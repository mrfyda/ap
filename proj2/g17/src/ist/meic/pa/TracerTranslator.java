package ist.meic.pa;

import javassist.CannotCompileException;
import javassist.ClassPool;
import javassist.NotFoundException;
import javassist.Translator;

public class TracerTranslator implements Translator {

    @Override
    public void start(ClassPool pool) throws NotFoundException, CannotCompileException {

    }

    @Override
    public void onLoad(ClassPool pool, String classname) throws NotFoundException, CannotCompileException {

    }
}
