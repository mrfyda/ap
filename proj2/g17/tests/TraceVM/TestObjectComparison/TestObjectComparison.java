import ist.meic.pa.Trace;

import java.util.HashMap;
import java.util.Map;

class TestObjectComparisonAux {

    Map m = new HashMap();

    public Object identity(Object o) {
        return o;
    }

    public void test() {
        Object o = new String("MyObj");

        identity(o);
        m.put(2, o);
        m.get(2);

        Trace.print(o);

        Object o1 = new String("MyObj");

        for (Object obj : m.values()) {
            System.out.println(obj);
        }

        Trace.print(o);
        Trace.print(o1);

    }
}

public class TestObjectComparison {
    public static void main(String args[]) {
        (new TestObjectComparisonAux()).test();
    }
}
