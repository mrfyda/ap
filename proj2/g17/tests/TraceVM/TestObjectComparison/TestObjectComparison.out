Tracing for MyObj
  <- java.lang.String(java.lang.String) on TestObjectComparison.java:15
  -> TestObjectComparisonAux.identity(java.lang.Object) on TestObjectComparison.java:17
  <- TestObjectComparisonAux.identity(java.lang.Object) on TestObjectComparison.java:17
  -> java.util.Map.put(java.lang.Object,java.lang.Object) on TestObjectComparison.java:18
  <- java.util.Map.get(java.lang.Object) on TestObjectComparison.java:19
  -> ist.meic.pa.Trace.print(java.lang.Object) on TestObjectComparison.java:21
MyObj
Tracing for MyObj
  <- java.lang.String(java.lang.String) on TestObjectComparison.java:15
  -> TestObjectComparisonAux.identity(java.lang.Object) on TestObjectComparison.java:17
  <- TestObjectComparisonAux.identity(java.lang.Object) on TestObjectComparison.java:17
  -> java.util.Map.put(java.lang.Object,java.lang.Object) on TestObjectComparison.java:18
  <- java.util.Map.get(java.lang.Object) on TestObjectComparison.java:19
  -> ist.meic.pa.Trace.print(java.lang.Object) on TestObjectComparison.java:21
  <- java.util.Iterator.next() on TestObjectComparison.java:25
  -> java.io.PrintStream.println(java.lang.Object) on TestObjectComparison.java:26
  -> ist.meic.pa.Trace.print(java.lang.Object) on TestObjectComparison.java:29
Tracing for MyObj
  <- java.lang.String(java.lang.String) on TestObjectComparison.java:23
  -> ist.meic.pa.Trace.print(java.lang.Object) on TestObjectComparison.java:30
